python/lsst/afw/cameraGeom/utils.py

```diff
@@ -336,64 +343,116 @@ def _prepareImage(self, ccd, im, binSize, allowRotate=True):
+                        ccd = im.getDetector() # possibly modified by assembleCcdTask
+                        e = None
+                        break
+                    except Exception as e:
+                        continue
+
+                if im:
+                    im = im.getMaskedImage().getImage()
+                else:
+                    raise e
 
             if e:
                 if self.verbose:
                     print("Reading %s: %s" % (ccd.getId(), e))
 
         if im is None:
-            return self._prepareImage(ccd, imageFactory(*bbox.getDimensions()), binSize)
+            return self._prepareImage(ccd, imageFactory(*bbox.getDimensions()), binSize), ccd
 
         if self.type == "raw":
+            allowRotate = True          # all other images were rotated by the ISR
             if hasattr(im, 'convertF'):
                 im = im.convertF()
-        else:
-            return self._prepareImage(ccd, im, binSize, allowRotate=False) # calexps were rotated by the ISR
 
-        ccdImage = im.Factory(bbox)
+        if self.callback:
+            try:
+                im = self.callback(im, ccd, imageSource=self)
+            except Exception, e:
+                if self.verbose:
+                    print("callback failed: %s" % e)
+                im = imageFactory(*bbox.getDimensions())
 
-        ampImages = []
-        for a in ccd:
-            bias = im[a.getRawHorizontalOverscanBBox()]
+        return self._prepareImage(ccd, im, binSize, allowRotate=allowRotate), ccd
+
+def rawCallback(im, ccd=None, imageSource=None,
+                correctGain=False, subtractBias=False, convertToFloat=False):
+    """!A callback function that may or may not subtract bias/correct gain/trim a raw image
+
+    @param im   An image of a chip, ready to be binned and maybe rotated.  Maybe an Exposure,
+                a MaskedImage or an Image
+    @param ccd  The Detector; if None assume that im is an exposure and extract it's Detector
+    @param imageSource  An object derived from FakeImageDataSource (typically ButlerImage)
+    @param correctGain  Correct each amplifier for its gain
+    @param subtractBias Subtract the bias from each amplifier
+    @param convertToFloat  Convert im to floating point if possible
+
+    N.b. if imageSource is derived from ButlerImage, imageSource.butler is available
+    """
+    if ccd is None:
+        ccd = cameraGeom.cast_Ccd(im.getDetector())
+    if hasattr(im, "getMaskedImage"):
+        im = im.getMaskedImage()
+    if convertToFloat and hasattr(im, "convertF"):
+        im = im.convertF()
+
+    isTrimmed = imageSource.isTrimmed
+    if isTrimmed:
+        bbox = ccd.getBBox()
+    else:
+        bbox = calcRawCcdBBox(ccd)
+
+    ccdImage = im.Factory(bbox)
+
+    ampImages = []
+    for a in ccd:
+        if isTrimmed:
             data = im[a.getRawDataBBox()]
+        else:
+            data = im
+
+        if subtractBias:
```

Comment:
Is this order of operations correct? (Note that not I'm asserting that it isn't, I'm just not sure). Also, should an option to correct for linearity be inserted here as well, or do you feel that's going too far? If so, then that factors in to the answer to the first question, as the definitions need to be consistent.